### PR TITLE
[OTBN] vertical attack initial work

### PR DESCRIPTION
**Pull request by m-temp**
This is a PR to bring together the works of Jade, Bilgiday and me

Co-authored-by: Jade Philipoom <jadep@google.com>
Signed-off-by: Michael Tempelmeier <michael.tempelmeier@gi-de.com>

Things to do:
- [x] Add TVLA support
    - [x] stimuli
    - [x] analysis script
- <del>[ ] Restructure code to support different code snippets</del>
    - <del>[ ] add select function to simple serial (python + c)</del>
    - <del>[ ] put "golden model" in a function to support different calculations</del>
- [x] Prepare and merge PR in opentitan repo ==> add binaries with specific hash id

### CODE REVIEW

File: cw/capture.py [hunk 1]
```diff
@@ -135,6 +135,7 @@ def capture_init(ctx, num_traces, plot_traces):
 
     # Key and plaintext generator
     ctx.obj.ktp = cw.ktp.Basic()
+    ctx.obj.ktp.fixed_key = False # if true, then key never updates and setting key_len has no effect
```
Comment (m-temp) [inline cw/capture.py:138]: This is a small change, but might have a huge impact!
By default the ktp.next always outputs the same key. I saw multiple workarounds throughout this project like using the kpt.plaintext_next as a random number generator for both key and plaintext, but I'm not sure if there are parts that amuse default behavior....

We should discuss this change.

File: cw/capture.py [hunk 2]
```diff
@@ -991,6 +992,182 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
     capture_end(ctx.obj.cfg)
 
 
+def capture_ecdsa_secret_keygen(ot, ktp, fw_bin, pll_frequency, capture_cfg):
```
Comment (vogelpi) [inline cw/capture.py:995]: I suggest to rename this to something more generic, e.g. `capture_otbn_vertical()`. The reason for this is that `keygen` is just one/the first code snippet we agreed to do a vertical analysis for. In the future, we most likely will analyze also other snippets for which the settings will be very similar (but pretty different from analyzing full ECDSA). In that sense vertical analysis of OTBN code snippets is far closer to analyzing AES/KMAC than ECDSA. We could even do batch captures here but shouldn't need streaming. Capture rate should be 200 MHz.

 What might change for other snippets and we should try to keep things flexible in this regard:
- number of samples per trace (but they should be far below the max 132k samples)
- offsets
- number of inputs and length of inputs
- output length

File: cw/capture.py [hunk 3]
```diff
@@ -991,6 +992,182 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
...
+
+    # OTBN operations are long. CW-Husky can store only 131070 samples
+    # in the non-stream mode. In case we want to collect more samples,
+    # we can run the operation multiple times with different adc_offset
+    # values.
+    # The trace from each run is kept in a section, and all sections are
+    # concatenated to create the final trace.
+    fifo_size = 131070
+    num_sections = capture_cfg["num_samples"] // fifo_size
+    print(f"num_sections = {num_sections}")
```
Comment (vogelpi) [inline cw/capture.py:1021]: I think this shouldn't be needed. Maybe it's best to start from the capture routine used for SHA3/KMAC for this. The code snippets should be small. If if they are 500 instructions and we oversample by 20x, the individual trace will just be 10000 samples.

File: cw/util/device.py [hunk 4]
```diff
@@ -76,8 +76,9 @@ def initialize_fpga(self, fpga, bitstream, pll_frequency):
         # bitstream.
         # Unfortunately, this doesn't seem to work for the CW310 yet,
         # see https://github.com/lowRISC/ot-sca/issues/48.
+        # This should work now. Issue is closed
```
Comment (vogelpi) [inline cw/util/device.py:79]: I think if the issues is resolved we should actually remove this reference here. :-)

File: cw/util/device.py [hunk 5]
```diff
@@ -76,8 +76,9 @@ def initialize_fpga(self, fpga, bitstream, pll_frequency):
...
         # TODO: We should have this in the CLI.
-        force_programming = False
+        force_programming = True
```
Comment (vogelpi) [inline cw/util/device.py:81]: IIRC this will cause the FPGA to always be re-programmed which we actually don't want no?

**Review by vogelpi — COMMENTED** [review]
Hi @m-temp ,
thanks for the PR!
I had a first brief look now. I think we should try to keep this capture setup as generic as possible because we're going to use it for other code snippets than keygen as well. Also, since the captures are short we don't need some of the infrastructure used in for ECDSA captures. The whole capture and analysis is much more similar to KMAC/AES. Maybe it's best to discuss this tomorrow, I've set up a little sync meeting.

File: cw/capture.py (hunk 1 above)
Comment (m-temp) [reply cw/capture.py]: This will be resolved in https://github.com/lowRISC/ot-sca/issues/116
For this PR we will use a workaround with backwards compatibility.

**Comment by m-temp** [comment]
I've added a first fixed-vs-random tvla version. There still a couple of todos like magic numbers etc.
Major points are:
 - finding good parameter sets (out of memory vs enough data)
 - add a batch version
 - restructure the code

**Comment by m-temp** [comment]
To expedite cooperation, I mark this PR as ready to merge

It adds:
  - basic fvsr support for otbn vertical
  - ctrl+c to store capture results
  - config outputs to tvla plots

A brief test confirmed that sha3, kmac, aes are not broken.

What is still missing:
  - doc for otbn vertical
  - a clear commit id for the binaries. (we will add this with a separate PR that also improves the software running on the device.)

File: cw/capture.py [hunk 6]
```diff
@@ -991,6 +1008,235 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
...
+    seed_fixed = ktp.next_key()
+    print(f'fixed seed = {seed_fixed.hex()}')
+    if len(seed_fixed) != seed_bytes:
+        raise ValueError(f'Fixed seed length is {len(seed_fixed)}, expected {seed_bytes}')
+
+    # Expected key is `seed mod n`, where n is the order of the curve and
+    # `seed` is interpreted as little-endian.
+    expected_fixed_key = int.from_bytes(seed_fixed, byteorder='little') % curve_order_n
+
+    # register ctrl-c handler to not loose already recorded traces if measurement is aborted
```
Comment (abdullahvarici) [inline cw/capture.py:1085]: Typo
```suggestion
    # register ctrl-c handler to not lose already recorded traces if measurement is aborted
```

File: cw/capture.py [hunk 7]
```diff
@@ -991,6 +1008,235 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
...
+
+            # Send the mask and start the keygen operation.
+            ot.target.simpleserial_write('k', mask)
+
+            # Wait until operation is done.
+            ret = ot.scope.capture(poll_done=True)
+            if ret:
+                raise RuntimeError('Timeout during capture')
+
+            # Check the number of cycles where the trigger signal was high.
```
Comment (abdullahvarici) [inline cw/capture.py:1128]: Should we compare it with an expected number instead of printing?

File: cw/capture.py [hunk 8]
```diff
@@ -991,6 +1008,235 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
...
+        if capture_cfg["masks_off"] is True:
+            # Use a constant mask for each trace
+            mask = bytearray(capture_cfg["plain_text_len_bytes"])  # all zeros
+        else:
+            # Generate a new random mask for each trace.
+            mask = ktp.next_text()
+        tqdm.write("Starting new trace....")
+        tqdm.write(f'mask   = {mask.hex()}')
+
+        if sample_fixed:
```
Comment (abdullahvarici) [inline cw/capture.py:1107]: The same code is shared in if and else part. Shouldn't we put it after the if else statement?

File: cw/capture.py [hunk 9]
```diff
@@ -991,6 +1008,235 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
...
+
+        sample_fixed = random.randint(0, 1)
+        if actual_key != expected_key:
+            raise RuntimeError('Bad generated key:\n'
+                               f'Expected: {hex(expected_key)}\n'
+                               f'Actual:   {hex(actual_key)}')
+
+        # Create a chipwhisperer trace object and save it to the project
+        # Args/fields of Trace object: waves, textin, textout, key
+        textout = share0 + share1  # concardinate bytearrays
```
Comment (abdullahvarici) [inline cw/capture.py:1196]: Typo
```suggestion
        textout = share0 + share1  # concatenate bytearrays
```

File: cw/capture.py [hunk 10]
```diff
@@ -991,6 +1008,235 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
     capture_end(ctx.obj.cfg)
 
 
+def capture_otbn_vertical(ot, ktp, fw_bin, pll_frequency, capture_cfg):
```
Comment (abdullahvarici) [inline cw/capture.py:1011]: I think we should either make the function name specific or move algorithm specific parts (generating input and comparing result) to specific functions.

File: cw/capture.py [hunk 11]
```diff
@@ -991,6 +1008,235 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
...
+@app_capture.command()
+def otbn_vertical(ctx: typer.Context,
+                  num_traces: int = opt_num_traces,
+                  plot_traces: int = opt_plot_traces):
+    """Capture ECDSA secret key generation traces."""
+    capture_init(ctx, num_traces, plot_traces)
+
+    # OTBN's public-key operations might not fit into the sample buffer of the scope
+    # These two parameters allows users to conrol the sampling frequency
+    #
```
Comment (abdullahvarici) [inline cw/capture.py:1213]: Do we need this empty comment line?

File: cw/capture.py [hunk 12]
```diff
@@ -991,6 +1008,235 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
...
+    # `adc_mul` affects the clock frequency (clock_freq = adc_mul * pll_freq)
+    #
```
Comment (abdullahvarici) [inline cw/capture.py:1215]: Do we need this empty comment line?

File: cw/tvla.py [hunk 13]
```diff
@@ -336,15 +336,18 @@ def run_tvla(ctx: typer.Context):
                     level=log.INFO,
                     force=True,)
 
-    if cfg["mode"] != "kmac" and cfg["mode"] != "aes" and cfg["mode"] != "sha3":
+    if (cfg["mode"] != "kmac" and cfg["mode"] != "aes" and cfg["mode"] != "sha3" and
+            cfg["mode"] != "sha3" and cfg["mode"] != "otbn"):
```
Comment (abdullahvarici) [inline cw/tvla.py:340]: Duplicate "sha3" check!

**Review by abdullahvarici — COMMENTED** [review]
Hello! Thanks for the PR. Sorry, I don't have any hardware to try it out for now. I think, in general it looks good. I've added a couple of feedbacks.

File: cw/capture.py (hunk 10 above)
Comment (m-temp) [reply cw/capture.py:1011]: This is a good point. I totally agree.
It was name specific at first, but then we discussed in one of the sca meetings that it would be good to eventually support multiple algorithms. While moving algorithmic specific parts to specific functions, I came to the conclusion that this should be done at a later point when we actually have multiple algorithms.

So what's the best way for now?
- Change the name (here and in the config files) back, merge the PR, and then change it again, when we actually have multiple algorithms?
- Leave it as it is, knowing that there will be multiple algorithms soon? 
- Move the algorithmic specific parts to specific functions? (actually, I don't want to do this in this PR, as the PR is already too large)

@vogelpi what do you think?

File: cw/capture.py [hunk 14]
```diff
@@ -60,6 +63,16 @@ def create_cw_segmented(ot, capture_cfg):
 }
 
 
+# handler for ctrl-c keyboard interrupts
+# saves capture project before exiting in case abort is intended
+# needs to be registered in every capture function before capture loop
+def abort_handler(project, sig, frame):
```
[the remark this answers is not in the document]
Comment (wettermo) [reply cw/capture.py:66]: Thanks, you're right this looks more consistent. I changed it.

**Comment by m-temp** [comment]
Thanks @abdullahvarici for your review and sry for all those typos in the code. I will address them.

File: cw/capture.py [hunk 15]
```diff
@@ -991,6 +1008,235 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
...
+    # For backwards compatibility this must be set in the capture config file.
+    # This is a workaroung for https://github.com/lowRISC/ot-sca/issues/116
+    if "use_fixed_key_iter" not in capture_cfg:
+        raise RuntimeError('use_fixed_key_iter not se set!')
+    if capture_cfg["use_fixed_key_iter"] is not False:
+        raise RuntimeError('use_fixed_key_iter must be set to false!')
+
+    # Currently, we need to reprogram the firmware before every sign operation
+    # TODO: Investigate if this is needed.
+    reset_firmware = False
```
[the remark this answers is not in the document]
Comment (m-temp) [reply cw/capture.py:1036]: Good point. I think @bilgiday investigated it and it is not needed anymore. Should we keep the option?
I'll change the comment, but removing the option should be done in a different PR

File: cw/capture.py [hunk 16]
```diff
@@ -60,6 +63,16 @@ def create_cw_segmented(ot, capture_cfg):
...
 
 
+def abort_handler(project, sig, frame):
+    """ Handler for ctrl-c keyboard interrupts:
+        Saves capture project before exiting, in case abort is intended.
+        Needs to be registered in every capture function before capture loop. """
+    if project is not None:
+        print("caught keyboard interrupt -> saving project (traces)...")
+        project.save()
+    sys.exit(0)
```
Comment (vogelpi) [inline cw/capture.py:73]: This has to go into a separate PR, because it's unrelated to OTBN vertical analysis. Instead it's useful for any captures and a great addition actually, thanks for adding this!

File: cw/capture.py (hunk 10 above)
Comment (vogelpi) [reply cw/capture.py:1011]: Let's stick to the generic name and move things around as we add support for other snippets. I also remember having discussed this before as @m-temp pointed out.

File: cw/capture.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (vogelpi) [reply cw/capture.py:1036]: Your proposal sounds good @m-temp.

File: cw/tvla_cfg_otbn.yaml [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+input_file: null
+output_file: null
+number_of_steps: 1
+ttest_step_file: null
+plot_figures: true
+general_test: true
+mode: otbn
+key_len_bytes: 40
+sample_start: null
+num_samples : 4000
```
Comment (vogelpi) [inline cw/tvla_cfg_otbn.yaml:18]: Missing newline at EOF

File: cw/tvla.py [hunk 18]
```diff
@@ -504,9 +535,12 @@ def run_tvla(ctx: typer.Context):
...
                 if project.waves[0].dtype == 'uint16':
                     traces = np.empty((num_traces, num_samples), dtype=np.uint16)
+                    log.info(f"Will use samples from {sample_start} to {sample_start+num_samples}")
                     for i_trace in range(num_traces):
-                        traces[i_trace] = project.waves[i_trace + trace_start]
-                else:
+                        traces[i_trace] = project.waves[i_trace +
+                                                        trace_start][sample_start:sample_start +
+                                                                     num_samples]
+                else:  # FIXME: The wave slicing is only tested for otbn
```
Comment (vogelpi) [inline cw/tvla.py:543]: This is a very nice feature. Thanks for adding this!

It will also be very useful for other targets. It's not expected to test this for all different targets (there are simply too many options and we have other priorities). However, I would appreciate if you changed the code such that there is simply an assertion to check that when sample_start and sample_end are set, also otbn is selected and provide a comment where this check is done. This way we have to change much less code later on when enabling it e.g. for AES or KMAC.

File: cw/tvla.py [hunk 19]
```diff
@@ -883,12 +883,16 @@ def run_tvla(ctx: typer.Context):
...
+                "TVLA of " + cfg["project_file"] + '\n' + "No. of traces: " + str(num_traces))
+            axs[0].plot(xaxs, single_trace, "k")
+            axs[0].set_ylabel("trace")
             for i_order in range(num_orders):
-                axs[1 + i_order].plot(ttest_trace[i_order, 0, 0], "k")
-                axs[1 + i_order].plot(c * threshold, "r")
-                axs[1 + i_order].plot(-threshold * c, "r")
+                axs[1 + i_order].plot(xaxs, ttest_trace[i_order, 0, 0], "k")
+                axs[1 + i_order].plot(xaxs, c * threshold, "r")
+                axs[1 + i_order].plot(xaxs, -threshold * c, "r")
```
Comment (vogelpi) [inline cw/tvla.py:895]: This commit needs to be merged with the one adding support for sample_start and num_samples. Also these two things should go into a separate PR.

File: cw/capture.py (hunk 11 above)
Comment (m-temp) [reply cw/capture.py:1213]: Not necessarily, but it's the same structure as in `ecdsa_simple` and `ecdsa_stream`.
Should I remove it there as well? Or better do that in a different PR?

File: cw/capture.py (hunk 12 above)
Comment (m-temp) [reply cw/capture.py:1215]: see above.

File: cw/tvla.py [hunk 20]
```diff
@@ -883,12 +883,16 @@ def run_tvla(ctx: typer.Context):
         else:
             #
             c = np.ones(num_samples)
+            xaxs = range(sample_start, sample_start + num_samples)
             fig, axs = plt.subplots(3, sharex=True)
-            axs[0].plot(single_trace, "k")
+            axs[0].set_title(
+                "TVLA of " + cfg["project_file"] + '\n' + "No. of traces: " + str(num_traces))
```
Comment (vogelpi) [inline cw/tvla.py:889]: I think just this line here actually belongs into this very commit.

**Review by vogelpi — COMMENTED** [review]
Overall this looks good and we should move forward. Many thanks for all the work you've put into this @m-temp and @wettermo .

In addition to the OTBN vertical support, this PR contains a few very nice additions that should be factored out into separate PRs. These additions are uncontroversial and are unrelated to OTBN vertical support.

**Comment by vogelpi** [comment]
Also updating the bitstreams and existing binaries can be factored out. Remember the goal is always to have as small PRs and commit as possible. This speeds up the review and allows us to progress faster as a team.

File: cw/capture.py [hunk 21]
```diff
@@ -135,7 +148,11 @@ def capture_init(ctx, num_traces, plot_traces):
 
     # Key and plaintext generator
     ctx.obj.ktp = cw.ktp.Basic()
-    ctx.obj.ktp.key_len = cfg["capture"]["key_len_bytes"]
+    # This is a workaroung for https://github.com/lowRISC/ot-sca/issues/116
+    if "use_fixed_key_iter" in cfg["capture"]:  # for backwarts compatibility
+        ctx.obj.ktp.fixed_key = cfg["capture"]["use_fixed_key_iter"]
+        # if fixed_key = true, then key never updates and setting...
+    ctx.obj.ktp.key_len = cfg["capture"]["key_len_bytes"]  # ...key_len has no effect
```
Comment (vrozic) [inline cw/capture.py:155]: nit: This is a bit strange way of writing comments. I would avoid having code and comment in the same line.

File: cw/capture.py (hunk 10 above)
Comment (vrozic) [reply cw/capture.py:1011]: I would leave it as is for now and do the name change in a fallow-up PR.
We are likely to have multiple capture functions related to otbn vertical attacks, so maybe we should discuss and agree on a naming strategy.

File: cw/capture.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (bilgiday) [reply cw/capture.py:1036]: With the new versions of the device-side firmware, we don't need it. I just kept it here as an option just in case. 

After we merge this PR, I can create a new one to clean `reset_firmware` options from the capture.py. Does that sound good?

File: cw/capture.py (hunk 11 above)
Comment (bilgiday) [reply cw/capture.py:1213]: I was keeping them for the readability. If it violates the styling guidelines, we can remove them with the `reset_firmware` PR proposed in [this comment](https://github.com/lowRISC/ot-sca/pull/115#discussion_r1118735037) above.

File: cw/capture.py (hunk 16 above)
Comment (bilgiday) [reply cw/capture.py:73]: Thanks @m-temp and @wettermo!

Comment (m-temp) [reply cw/capture.py:73]: I will remove that from the current PR

File: cw/capture.py (hunk 7 above)
Comment (m-temp) [reply cw/capture.py:1128]: We could do that, however the actual number of cycles varies by `+-2`  cycles.
If needed, let's do that in another PR.

File: cw/tvla.py (hunk 18 above)
Comment (m-temp) [reply cw/tvla.py]: wilco

File: cw/tvla.py (hunk 19 above)
Comment (m-temp) [reply cw/tvla.py]: wilco

File: cw/tvla.py (hunk 20 above)
Comment (m-temp) [reply cw/tvla.py]: Hmm let's put it also in the other PR, where we change the plotting

**Comment by m-temp** [comment]
Thx all for the reviews. I changed it accordingly and will file a couple of PRs for the remaining features.

File: cw/capture.py [hunk 22]
```diff
@@ -991,6 +1008,235 @@ def kmac_fvsr_key_batch(ctx: typer.Context,
...
+    if capture_cfg["num_samples"] > fifo_size:
+        raise RuntimeError('Current setup only supports up to 130k samples')
+
+    # Be sure we don't use the stream mode
+    if ot.scope._is_husky:
+        ot.scope.adc.stream_mode = False
+        ot.scope.adc.bits_per_sample = 12
+        ot.scope.adc.samples = capture_cfg["num_samples"]
+    else:
+        # TODO: Add cw-lite support
```
[the remark this answers is not in the document]
Comment (vogelpi) [reply cw/capture.py:1037]: Hm, I am not sure I understand your comment. I think we don't want to have a TODO to Add CW-Lite support, because we're never going to add it but to error out if CW-Lite is detected is the best thing to do here.

**Review by vogelpi — APPROVED** [review]
Thanks for addressing the comments. This looks good!

Two minor nits:
1. There is still a TODO for the CW-Lite, but we won't add support for this. Having the script to error out for something else than Husky still desirable though.
2. The PR also adds a new bitstream in the first commit. Is this really needed? We wanted to add the new bistream with the different URND behavior in a separate PR I believe.

**Comment by m-temp** [comment]
>     1. There is still a TODO for the CW-Lite, but we won't add support for this. Having the script to error out for something else than Husky still desirable though.
Sry. I missed that. This was a leftover from Jade. I'll remove the comment, but keep the if-else-branch 

>     2. The PR also adds a new bitstream in the first commit. Is this really needed? We wanted to add the new bistream with the different URND behavior in a separate PR I believe.
Ah okay, I thought this was only related to the second bins, but you are right, I delete the old bins and file a separate PR for the new ones.
